firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigAutoFetch.java

```diff
@@ -194,35 +200,78 @@ public void run() {
 
-    // fetchAttemptNumber is calculated by subtracting current attempts from the max number of
-    // possible attempts.
     Task<ConfigFetchHandler.FetchResponse> fetchTask =
         configFetchHandler.fetchNowWithTypeAndAttemptNumber(
-            ConfigFetchHandler.FetchType.REALTIME, MAXIMUM_FETCH_ATTEMPTS - currentAttempts);
-    fetchTask.onSuccessTask(
-        (fetchResponse) -> {
-          long newTemplateVersion = 0;
-          if (fetchResponse.getFetchedConfigs() != null) {
-            newTemplateVersion = fetchResponse.getFetchedConfigs().getTemplateVersionNumber();
-          } else if (fetchResponse.getStatus()
-              == ConfigFetchHandler.FetchResponse.Status.BACKEND_HAS_NO_UPDATES) {
-            newTemplateVersion = targetVersion;
-          }
+            ConfigFetchHandler.FetchType.REALTIME, currentAttemptNumber);
+    Task<ConfigContainer> activatedConfigsTask = activatedCache.get();
 
-          if (newTemplateVersion >= targetVersion) {
-            executeAllListenerCallbacks();
-          } else {
-            Log.d(
-                TAG,
-                "Fetched template version is the same as SDK's current version."
-                    + " Retrying fetch.");
-            // Continue fetching until template version number if greater then current.
-            autoFetch(currentAttempts, targetVersion);
-          }
-          return Tasks.forResult(null);
-        });
+    return Tasks.whenAllComplete(fetchTask, activatedConfigsTask)
+        .continueWithTask(
+            scheduledExecutorService,
+            (listOfUnusedCompletedTasks) -> {
+              if (!fetchTask.isSuccessful()) {
+                return Tasks.forException(
+                    new FirebaseRemoteConfigClientException(
+                        "Failed to auto-fetch config update.", fetchTask.getException()));
+              }
+
+              if (!activatedConfigsTask.isSuccessful()) {
+                return Tasks.forException(
+                    new FirebaseRemoteConfigClientException(
+                        "Failed to get activated config for auto-fetch",
+                        activatedConfigsTask.getException()));
+              }
+
+              ConfigFetchHandler.FetchResponse fetchResponse = fetchTask.getResult();
+              ConfigContainer activatedConfigs = activatedConfigsTask.getResult();
+
+              if (!fetchResponseIsUpToDate(fetchResponse, targetVersion)) {
+                Log.d(
+                    TAG,
+                    "Fetched template version is the same as SDK's current version."
+                        + " Retrying fetch.");
+                // Continue fetching until template version number is greater then current.
+                autoFetch(remainingAttemptsAfterFetch, targetVersion);
+                return Tasks.forResult(null);
+              }
+
+              if (fetchResponse.getFetchedConfigs() == null) {
```

Comment:
It looked like getFetchedConfigs() might return a value when the state is BACKEND_HAS_NO_UPDATES, since it's constructed with one here: https://github.com/firebase/firebase-android-sdk/blob/993673edf6a6e1c6dd159238fbab73c0a8d2f84f/firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigFetchHandler.java#L650.

It also seemed safer to check for null before getting the value later on in this method.

Reply:
I don't think there is a scenario anymore where the response returns a null fetchedConfigs, due to the change where we added the template version to the NO_CHANGE response. Plus in the fetchResponseisUpToDate method it checks the fetchedConfig for null and the only way to return true for when the fetchedConfig is null is if the status is BACKEND_HAS_NO_UPDATES

Reply:
Wondered if we could remove this check?

Reply:
Good point it likely isn't necessary. I'd keep it for now since `getFetchedConfigs()` isn't annotated `@Nonnull` so it _could_ be null and checking for `null` keeps with the assumptions that the rest of the SDK is written with right now. We might change that separately later.